### PR TITLE
Import only named params into FIPS module

### LINKED ISSUE #17978: ecparam -check in fips mode
Current master

The command
```./util/wrap.pl -fips `pwd`/apps/openssl ecparam -out 384.pem -name secp384r1 -param_enc explicit```
works normally.

The command
```./util/wrap.pl -fips `pwd`/apps/openssl ecparam -check -in 384.pem```
gives the following output:
```
Could not read params etc. of EC parameters from 384.pem
4027A373D87F0000:error:07800081:common libcrypto routines:get_ptr_internal:param of incompatible type:crypto/params.c:1416:
4027A373D87F0000:error:07800081:common libcrypto routines:get_ptr_internal:param of incompatible type:crypto/params.c:1416:
4027A373D87F0000:error:07800081:common libcrypto routines:get_ptr_internal:param of incompatible type:crypto/params.c:1416:
4027A373D87F0000:error:1608010C:STORE routines:ossl_store_handle_load_result:unsupported:crypto/store/store_result.c:151:
4027A373D87F0000:error:07800081:common libcrypto routines:get_ptr_internal:param of incompatible type:crypto/params.c:1416:
4027A373D87F0000:error:07800081:common libcrypto routines:get_ptr_internal:param of incompatible type:crypto/params.c:1416:
```
Is it desired (and documented) behaviour?

### PR DESCRIPTION
**Pull request by t8m**
Fixes #17978

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Review by paulidale — APPROVED** [review]

File: crypto/ec/ec_lib.c [hunk 1]
```diff
@@ -1559,6 +1566,9 @@ EC_GROUP *EC_GROUP_new_from_params(const OSSL_PARAM params[],
         }
         return group;
     }
+#ifdef FIPS_MODULE
+    return NULL;
```
Comment (romen) [inline crypto/ec/ec_lib.c:1570]: Should we raise an error here about unsupported explicit parameters?

Comment (t8m) [reply crypto/ec/ec_lib.c:1570]: Good point, I'll add it.

**Comment by t8m** [comment]
@paulidale still OK?

**Review by romen — APPROVED** [review]
LGTM assuming tests pass

**Comment by t8m** [comment]
@romen but we've already discussed that on OTC that we do not want this in 3.0. But I do not mind discussing it again.

**Comment by romen** [comment]
> @romen but we've already discussed that on OTC that we do not want this in 3.0. But I do not mind discussing it again.

We had consensus that the 2 issues (the encoder/decoder one and the FIPS module one) had to be addressed separately, as the first was unanimously seen as a bug fix.
Now that this PR properly defines what we intend to fix for master, I would like to call the attention of the OTC to this again.

**Comment by paulidale** [comment]
A hypothetical question:

What happens when another validated FIPS provider wants to send our FIPS provider a curve via key management.
It might send a curve as explicit parameters.  The base provider won't be able to translate the curve.

I'm not sure if this is even done, let alone common.

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
A hypothetical question:

What happens when another validated FIPS provider wants to send our FIPS provider a curve via key management.
It might send a curve as explicit parameters.  The base provider won't be able to translate the curve.

I'm not sure if this is even done, let alone common.

**Comment by t8m** [comment]
> What happens when another validated FIPS provider wants to send our FIPS provider a curve via key management.
> It might send a curve as explicit parameters. The base provider won't be able to translate the curve.

Of course then the import into our FIPS provider fails. I do not think we should care about this scenario.

**Comment by romen** [comment]
> A hypothetical question:
> 
> What happens when another validated FIPS provider wants to send our FIPS provider a curve via key management. It might send a curve as explicit parameters. The base provider won't be able to translate the curve.
> 
> I'm not sure if this is even done, let alone common.

If at the FIPS boundary we should accept or not explicit parameters (for the FIPS-approved named curves) should really depend on what are the FIPS requirements are about this: if only FIPS "named curves" are supposed to be supported within the FIPS module, I would see the reject as expected (and thus this would be a bugfix). I'd like to hear @slontis feedback on this!

Even if we already have checks inside the FIPS module that make ECDH and ECDSA operations on arbitrary curves fail, allowing key objects constructed on _mostly unchecked_ external input to exist within the boundary of the FIPS module can still be a security risk, as we have seen recently.
While there are arguments to keep supporting explicit parameters outside the FIPS boundary (e.g., to support standardized curves that have not been assigned an OID), when it comes to entering the FIPS boundary, arbitrary curves should be rejected as soon as possible.
If we deem that not doing so is a bug in the 3.0 FIPS module, we could:
- backport this PR (which disables explicit parameters altogether within FIPS, _even those matching FIPS curves_)
- have a separate PR that alters https://github.com/openssl/openssl/blob/15dd39411f47ce5dc104a578530d50e9335d160d/providers/implementations/keymgmt/ec_kmgmt.c#L368-L423 so in `common_import()`, after returning from `ossl_ec_group_fromdata()`,  we check within a `#ifdef FIPS_MODE` block, if we matched one of the FIPS supported named curves (either inline, or with a function conceptually similar to `common_check_sm2()`).

The second approach would allow us to reject arbitrary curves, but still accept FIPS curves encoded with explicit parameters that match the ones in our internal tables. Notice though that in this case, if the FIPS provider is receiving the explicit parameters via key management from another provider, nothing would prevent reaching vulnerable points in the code that is parsing the parameters, e.g., reproducing the recent DoS CVE fixed in the latest 3.0 release, by sending a well-chosen combination of `p` and compressed-form generator point.

**Comment by t8m** [comment]
OTC approved this for both 3.0 and master branches. Merging.

**Comment by t8m** [comment]
Merged to master and 3.0 branches. Thank you for your reviewes.

**Pull request closed without merging**